api/v1alpha1/thanosruler_types.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ThanosRulerSpec defines the desired state of ThanosRuler
+type ThanosRulerSpec struct {
+	CommonThanosFields `json:",inline"`
+	// Replicas is the number of Ruler replicas.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Required
+	Replicas int32 `json:"replicas,omitempty"`
+	// By default, the operator will add all discoverable Queriers to the Ruler,
+	// if they have query labels. You can optionally choose to override default
+	// Query selector labels, to select a subset of Queries to query.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={matchLabels:{"operator.thanos.io/query-api": "true"}}
+	QueryLabelSelector *metav1.LabelSelector `json:"queryLabelSelector,omitempty"`
+	// ObjectStorageConfig is the secret that contains the object storage configuration for Ruler to upload blocks.
+	// +kubebuilder:validation:Required
+	ObjectStorageConfig ObjectStorageConfig `json:"defaultObjectStorageConfig,omitempty"`
+	// The operator will mount all ConfigMaps with the label operator.thanos.io/rule-file=true
+	// to the Ruler as rule file. You can optionally choose to override default Rule Config selector labels,
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={matchLabels:{"operator.thanos.io/rule-file": "true"}}
+	RuleConfigSelector *metav1.LabelSelector `json:"ruleConfigSelector,omitempty"`
+	// AlertmanagerURL is the URL of the Alertmanager to which the Ruler will send alerts.
+	// +kubebuilder:validation:Required
+	AlertmanagerURL string `json:"alertmanagerURL,omitempty"`
```

Comment:
I wonder if we should clarify that this can use dns service discovery? 

I also wonder based on the following text `The scheme should not be empty
                                 e.g `http` might be used. The scheme may be
                                 prefixed with 'dns+' or 'dnssrv+' to detect
                                 Alertmanager IPs through respective DNS
                                 lookups.` if we should enforce that pattern on this string (for scheme)